source/Crew.cpp

```diff
@@ -0,0 +1,232 @@
+	dailySalary = 100;
+	minimumPerShip = 0;
+	populationPerOccurrence = 0;
+	
+	for(const DataNode &child : node)
+	{
+		if(child.Size() >= 2)
+		{
+			if(child.Token(0) == "daily salary")
+				dailySalary = child.Value(1);
+			else if(child.Token(0) == "minimum per ship")
+				minimumPerShip = child.Value(1);
+			else if(child.Token(0) == "population per occurrence")
+				populationPerOccurrence = child.Value(1);
+			else if(child.Token(0) == "name")
+				name = child.Token(1);
+			else
+				child.PrintTrace("Skipping unrecognized attribute:");
+		}
+		else if(child.Token(0) == "avoids escorts")
+			avoidsEscorts = true;
+		else if(child.Token(0) == "avoids flagship")
+			avoidsFlagship = true;
+		else if(child.Token(0) == "pay salary while parked")
+			isPaidSalaryWhileParked = true;
+		else
+			child.PrintTrace("Skipping incomplete attribute:");
+	}
+}
+
+
+
+int64_t Crew::CalculateSalaries(
+	const vector<shared_ptr<Ship>> ships,
+	const bool includeExtras
+)
+{
+	int64_t totalSalaries = 0;
+	
+	for(const shared_ptr<Ship> &ship : ships)
+	{
+		totalSalaries += Crew::SalariesForShip(
+			ship,
+			// Determine whether if the current ship is the flagship
+			ship->Name() == ships.front()->Name(),
+			includeExtras
+		);
+	}
+	
+	return totalSalaries;
+}
+
+
+
+int64_t Crew::CostOfExtraCrew(const vector<shared_ptr<Ship>> ships)
+{
+	// Calculate with and without extras and return the difference.
+	return Crew::CalculateSalaries(ships, true)
+		- Crew::CalculateSalaries(ships, false);
+}
+
+
+
+int64_t Crew::NumberOnShip(
+	const Crew crew,
+	const shared_ptr<Ship> ship,
+	const bool isFlagship,
+	const bool includeExtras
+)
+{
+	int64_t count = 0;
+	
+	// If this is the flagship, check if this crew avoids the flagship.
+	if(isFlagship && crew.AvoidsFlagship())
+		return count;
+	// If this is an escort, check if this crew avoids escorts.
+	if(!isFlagship && crew.AvoidsEscorts())
+		return count;
+	
+	const int64_t countableCrewMembers = includeExtras
+		? ship->Crew()
+		: ship->RequiredCrew();
+	
+	// Apply the per-ship minimum.
+	 count = min(crew.MinimumPerShip(), countableCrewMembers);
+	
+	// Prevent division by zero so that the universe doesn't implode.
+	if(crew.PopulationPerOccurrence())
+	{
+		// Figure out how many of this kind of crew we have, by population.
+		count = max(
+			count,
+			countableCrewMembers / crew.PopulationPerOccurrence()
+		);
+	}
+	
+	return count;
+}
+
+
+
+int64_t Crew::SalariesForShip(
+	const shared_ptr<Ship> ship,
+	const bool isFlagship,
+	const bool includeExtras
+)
+{
+	// We don't need to pay dead people.
+	if(ship->IsDestroyed())
+		return 0;
+	
+	const Set<Crew> crews = GameData::Crews();
+	const Crew *defaultCrew = crews.Find("default");
+	
+	int64_t salariesForShip = 0;
+	int64_t specialCrewMembers = 0;
+	
+	// Add up the salaries for all of the special crew members
+	for(const pair<const string, Crew>& crewPair : crews)
+	{
+		// Skip the default crew members.
+		if(crewPair.first == "default")
```

Comment:
Is there a way to avoid needing to treat one member of the collection differently? Perhaps GameData holds a crew class that has the default type's values, and then that class has a collection of the other crewTypes?

Then you'd be able to have each member of the collection be handled in the same way, and the remaining needed crew are handled by the default type config.

Reply:
The reason they're handled differently is that there isn't a straight-forward way to figure out how many there are on the ship. They're the crew type that fills in the blanks after we've figured out all the special roles, so we have to find all the special roles first.

I wanted us to be able to use the data files to specify the behaviour for the default crew as well. Would your suggestion get us that feature, or would they be defined at the code level?

Reply:
My suggestion is a bit different design - only one `CrewEconomy` class for GameData to manage, and the datafile would look something like
```
"crew economy"
    salary 100
    type "Pilot"
        salary 125
        "escort only"
        minimum 1
    type "Officer"
        salary 110
        "crew fraction" .1
    type "XO"
        salary 175
        "always paid"
    ...
```

Note that one could easily enter a scenario where a ship cannot satisfy the minimum values of all crewTypes due to lack of bunks.

Reply:
Interesting idea. How well would this play with plugins overriding the crew definitions? Consider the following scenario:

1. Base game data defines a set of crews.

2. Plugin wants to:
	a. Define one additional type of crew
	b. Override one of the base game's crews

The plugin doesn't want to have to redefine the whole tree because they want to allow the base game to keep updating it.

What would the plugin put in their `plugin/crew.txt`?

Reply:
You would implement `add / remove` syntax for collection variables.

Reply:
I gave this some thought and had a look at the `add / remove` examples throughout the codebase.

Honestly, the single-node solution seems needlessly complex for what we're trying to achieve here. I'd prefer to stick with one node per crew type.

Reply:
I agree with tehhowch about not having an element in the collection treated differently.
But I like the original implementation for the datafiles from lukearndt better.

Can we change the code to automatically fill up open ship positions with the cheapest crew member available? So no special and regular crewmembers, but threat all crew as special, then fill all the minimum numbers and then just fill any left-over positions with the cheapest member possible.

Reply:
That's a neat idea.